File: app/app-swap/src/ui.c
#include <stdio.h>
#include <stdlib.h>

#include "sdk.h"
#include "ux/glyphs.h"
#include "ux/ux.h"

#include "swap.h"

static char g_ux_send_amount[50];
static char g_ux_get_amount[50];
static char g_ux_fees[50];

static struct ux_item_s ux_opensea[] = {
    { &C_icon_eye, "Review", "transaction", UX_ACTION_NONE },
    { NULL, "Send", g_ux_send_amount, UX_ACTION_NONE },
    { NULL, "Get", g_ux_get_amount, UX_ACTION_NONE },
    { NULL, "Fees", g_ux_fees, UX_ACTION_NONE },
    { &C_icon_validate_14, "Accept", "and send", UX_ACTION_VALIDATE },
    { &C_icon_crossmark, "Reject", NULL, UX_ACTION_REJECT },
};

/* return true if the user approved the tx, false otherwise */
bool ui_sign_tx_validation(const char *send_amount, const char *get_amount, const char *fees)
{
    strncpy(g_ux_send_amount, send_amount, sizeof(g_ux_send_amount));
    strncpy(g_ux_get_amount, get_amount, sizeof(g_ux_get_amount));
    strncpy(g_ux_fees, fees, sizeof(g_ux_fees));

    bool app_loading = app_loading_stop();

    bool validated = ux_validate(ux_opensea, ARRAY_SIZE(ux_opensea));

    if (app_loading) {
        app_loading_start("Signing transaction...");
    } else {
        ux_idle();
    }

    return validated;
}

File: app/app-rust/src/ui.rs
use sdk::glyphs::{ICON_CROSSMARK, ICON_EYE, ICON_VALIDATE};
use sdk::ux::*;

pub fn sign_tx_validation(send_amount: &str, get_amount: &str, fees: &str) -> bool {
    let swap_ui: [UxItem; 6] = [
        UxItem {
            icon: Some(&ICON_EYE),
            line1: "Review",
            line2: Some("Transaction"),
            action: UxAction::None,
        },
        UxItem {
            icon: None,
            line1: "Send",
            line2: Some(send_amount),
            action: UxAction::None,
        },
        UxItem {
            icon: None,
            line1: "Get",
            line2: Some(get_amount),
            action: UxAction::None,
        },
        UxItem {
            icon: None,
            line1: "Fees",
            line2: Some(fees),
            action: UxAction::None,
        },
        UxItem {
            icon: Some(&ICON_VALIDATE),
            line1: "Accept",
            line2: Some("and send"),
            action: UxAction::Validate,
        },
        UxItem {
            icon: Some(&ICON_CROSSMARK),
            line1: "Reject",
            line2: None,
            action: UxAction::Reject,
        },
    ];

    app_loading_stop();
    ux_validate(&swap_ui)
}
